purge/purge.py

```diff
@@ -0,0 +1,330 @@
+
+    with psycopg2.connect(db_uri) as db, db.cursor() as cur:
+
+        # set user as admin in database if needed
+        cur.execute(
+            'SELECT admin FROM users WHERE name = %s ;',
+            (admin_user,),
+        )
+        if not cur.rowcount:
+            raise RuntimeError(f'User {admin_user!r} not found')
+        is_admin, = cur.fetchone()
+        if admin_set and not is_admin:
+            cur.execute(
+                'UPDATE users SET admin=1 WHERE name = %s ;',
+                (admin_user,),
+            )
+            db.commit()
+        elif not is_admin:
+            raise RuntimeError(f'User {admin_user!r} is not an admin. See --admin-set option')
+
+        purges = dict()
+        def wait_and_purge_room(room_id=None, event_id=None):
+            """ Wait for available slots in parallel_purges and purge room
+
+            If room_id is None, just wait for current purges to complete and return
+            If event_id is None, purge all events in room
+            """
+            while len(purges) >= (parallel_purges if room_id else 1):
+                # wait and clear completed purges
+                time.sleep(1)
+                for _room_id, purge_id in list(purges.items()):
+                    response = session.get(
+                        urljoin(
+                            server,
+                            '/_matrix/client/r0/admin/purge_history_status/' + quote(purge_id),
+                        ),
+                        params={'access_token': admin_access_token},
+                    )
+                    if response.status_code != 200 or response.json().get('status') != 'active':
+                        click.secho(f'Finished purge: room {_room_id!r}, purge {purge_id!r}')
+                        purges.pop(_room_id)
+
+            if not room_id:
+                return
+
+            body = {'delete_local_events': True}
+            if event_id:
+                body['purge_up_to_event_id'] = event_id
+            else:
+                body['purge_up_to_ts'] = int(time.time() * 1000)
+            response = session.post(
+                urljoin(server, '/_matrix/client/r0/admin/purge_history/' + quote(room_id)),
+                params={'access_token': admin_access_token},
+                json=body,
+            )
+            if response.status_code == 200:
+                purge_id = response.json()['purge_id']
+                purges[room_id] = purge_id
+                return purge_id
+
+        if not keep_newer and not keep_min_msgs:
+            click.confirm(
+                'No --keep-newer nor --keep-min-msgs option provided. Purge all history?',
+                abort=True,
+            )
+
+        ts_ms = None
+        if keep_newer:
+            ts = datetime.datetime.now() - datetime.timedelta(keep_newer)
+            ts_ms = int(ts.timestamp() * 1000)
+
+        cur.execute('SELECT room_id FROM rooms ;')
+        all_rooms = {row for row, in cur}
```

Comment:
That comma seems superfluous.

Reply:
It's not. DBM rows always return a `tuple`, even if it only has one column, in this case, each row will be like `('<room_id>',)`, the comma there extracts the room_id. `row[0]` would work as well, but I think the comma syntax is clearer

Reply:
Ah unpacking a 1-element tuple, right.